### DISCUSSION TITLE
Auto-installation bash script feedback

### DISCUSSION THREAD

**Discussion by JimJamJamie** [azerothcore/azerothcore-wotlk#16583] [General]
Hello,

I would like to provide some feeback from testing the installer script on my Ubuntu 22.04 machine. First of all, thanks for your efforts on AzerothCore, I'm excited to spin up a server for myself, and this script is very handy. 

There were two errors encountered when running the First Installation step using the /apps/installer/main.sh script. 

Version below:
```
* AzerothCore revision            : 7945e11557a7+ 2023-05-18 15:34:50 -0300 (master branch)
* AzerothCore buildtype           : Release
```

The first error is as follows:
```
E: The repository 'https://ppa.launchpadcontent.net/mhier/libboost-latest/ubuntu jammy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```
I resolved this issue by installing the `libboost-all-dev` package from the Ubuntu repository: `sudo apt install libboost-all-dev`.

The second error:
```
CMake Error at deps/readline/CMakeLists.txt:25 (message):
  ** Readline library not found!

  ** Your distro may provide a binary for Readline e.g.  for ubuntu try
  apt-get install libreadline5-dev
```

Was resolved by installing the `libreadline-dev` package from the Ubuntu repository: `sudo apt install libreadline-dev`.

Once these libraries were manually installed,  the installation script completed succesfully and compiled. It would be fantastic (especially for less technical users) if the script could be modified to eliminate these minor issues, as it would make installation as simple as 'press 1 for success'.

If there is anything else I can provide, please ask away and I will endeavour to do so.

System configuration:
![image](https://github.com/azerothcore/azerothcore-wotlk/assets/21192547/c019873e-b676-4475-a795-46800912bc28)

Thanks

**Comment by JimJamJamie** [comment]
To add to this, installing the dependencies listed here: https://www.azerothcore.org/wiki/linux-requirements resolved the issues:

`sudo apt-get update && sudo apt-get install git cmake make gcc g++ clang libmysqlclient-dev libssl-dev libbz2-dev libreadline-dev libncurses-dev mysql-server libboost-all-dev`

Perhaps this list of packages could be integrated into the auto-install script for steps 1 and 2.

Another possible addition would be a 'configure database' option, which would execute this script: https://github.com/azerothcore/azerothcore-wotlk/blob/master/data/sql/create/create_mysql.sql

If this was included, then the install script would take the user through the complete installation process and have a running world/auth server simply by selecting the necessary menu options.

I would also consider adding the admin account by default as documented here: https://www.azerothcore.org/wiki/creating-accounts so that the user could log in right away post-install (perhaps with a notice to change the default password).

**Comment by JimJamJamie** [comment]
I set up an install on a minimal Debian 12 install today for further investigation into the install scripts. The goal of this discussion is to minimise the manual work involved during the install process of AzerothCore to make it as straightforward for the end user as possible.

## Dependencies required to allow the compilation to run successfully

- cmake (--no-install-recommends works fine, saves disk space)
- make
- clang (--no-install-recommends works fine, saves disk space)
- mariadb-server
- default-libmysqlclient-dev
- libboost-system-dev
- libboost-filesystem-dev
- libboost-program-options-dev
- libboost-iostreams-dev
- libreadline-dev (pulls in libncurses-dev)

## Notable omissions
The build process completed without the below packages present on my system, but are listed as dependencies in `\apps\installer\includes\os_configs\debian.sh`.

- gdbserver
- g++
- ccache

I'd be interested to know what function they are supposed to provide.

## No `deno`
The `deno` package is required only for docker stuff, but the `denoInstall` function (found in `bash_shared/includes.sh`) is called every time you run the `acore/installer.sh` script. It should be moved to where it is only called when the user selects docker-related options - automatically installing an unnecessary package is not good behaviour.

## Database package choice
On the subject of the debian OS config script, it is hardcoded to use MySQL instead of MariaDB. As per my package list above, MariaDB functions correctly as a drop-in replacement. Presenting the user with a choice of db to install would be ideal.

## Automate database setup
Add another menu option to the apps/installer/main.sh script, before option 11 (run-worldserver) and logically perhaps before option 10 (client-data) to run the below `database_setup` script (or something equivalent to it). This script executes the `create_mysql` script located in `data/sql/create`. This configures the database ready for running worldserver and authserver without the user manually logging in or typing commands. It is an automation of the "Set-up the database" step from the [installation guide](https://www.azerothcore.org/wiki/ac-dashboard-core-installation).

### database_setup.sh
```
#!/bin/bash
echo Creating user: \'acore\'  @ \'localhost\', with password: \'acore\'\

echo Creating databases: \'acore_world\', \'acore_characters\', \'acore_auth\'
sudo mysql -u root < data/sql/create/create_mysql.sql
```

## Automate creation of authserver and worldserver config files
I'm not sure which script to add it to, but I see no reason to require the user to manually copy worldserver and authserver.conf. Please add the following commands somewhere appropriate in the install process:

```
cp env/dist/etc/authserver.conf.dist env/dist/etc/authserver.conf
cp env/dist/etc/worldserver.conf.dist env/dist/etc/worldserver.conf
```

## Add host IP to realmlist table automagically
Populate the server LAN IP address in the acore_auth.realmlist.localAddress field for LAN connectivity without the user running sql insert statements.

See below for a rough linux implementation:

```
#!/bin/bash
ips=$(ip -o -4 addr show up primary scope global |
      while read -r num dev fam addr rest; do echo ${addr%/*}; done)

mysql -u acore -p -D acore_auth -e "UPDATE realmlist SET address = '${ips}' WHERE id = 1;"
```

## Built-in admin account

To provide an administrator account out of the box, the below SQL statements could be executed during installation:

```
INSERT INTO `account` (`id`,`username`,`salt`,`verifier`) VALUES (1,'ADMIN','','');

INSERT INTO `account_access` (`id`,`gmlevel`,`RealmID`) VALUES (1,3,-1);
```

Caveats: the user would still have to set a password using `account set password` (technically could use pre-computed values for `salt` and `verifier`, but I have a feeling this would be a bad idea), and I don't think the above method will be as reliable as just using `account create`.

## To Do

Create an example of the additional menu option to be added to `installer.sh`.

Create an example of the menu choice for picking a database package.

## Final thoughts
If these changes were made to the install script, there would be no bash or sql commands required of the user. They would simply select the required menu options and be left with a running server which they could log into.

Feedback welcome.

**Comment by michaeldelago** [comment]
Hey @JimJamJamie, I agree with most of your points. 

> Notable omissions

The most interesting one of these is `ccache`, which is caching for the compiler. `g++` is the gnu compiler for c++ and `gdbserver` is a server for the GNU debugger

> No `deno`

Yeah. Deno doesn't add anything and it actually is running shell under the hood. This means that it's going from shell -> typescript -> shell. Additionally, Deno doesnt support an arm64 build for MacOS so it's a problem there as well. 

I have a PR in to remove deno:

https://github.com/azerothcore/azerothcore-wotlk/pull/16934

> Database package choice

I agree. I don't think the dashboard should manage packages for the user

> Automate database setup

I'd prefer the script do less rather than more, though I do agree there should be very clear, simple, and easy documentation on what to do to bootstrap the database. Note as well that there is the "dbimport" tool that can do these, it acts like a database migration tool similar to dbmate or sqlx for rust.

> Automate creation of authserver and worldserver config files

Yeah. I think that it should just not fail (or even log a warning) if there isn't a `conf` file.

> Add host IP to realmlist table automagically

I'd be most interested in this as a config parameter for the world server. The easiest thing I can think of is to have an environment variable and if it's not an empty string, insert that value into the database.

> Built-in admin account

yes, 1000x. I really don't like the hard requirement on using the console to get started

> To Do
> Create an example of the additional menu option to be added to installer.sh.
> Create an example of the menu choice for picking a database package.

I'd rather the scripts do less rather than more. They jump around quite a bit, make some large assumptions about the users machine, and are at times difficult to maintain. They are quite complex and have a large amount of functions that feels like bloat more often than not.
